File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.java
class BSTIterator{

    Stack<TreeNode> stack;
    // reverse -> true -> before
    // reverse -> false -> after
    boolean reverse = true;

    public BSTIterator(TreeNode root, boolean isReverse){
        stack = new Stack<>();
        reverse = isReverse;
        pushAll(root);
    }
    
    public boolean hasNext(){
        return stack.size() != 0;
    }
    
    public int next(){
        
        TreeNode tmpNode = stack.pop();
        
        if(reverse == true){ // before
            pushAll(tmpNode.left);
        }
        else{
            pushAll(tmpNode.right);
        }
        
        return tmpNode.val;
    }
    
    private void pushAll(TreeNode node){
        
        while(node != null){
            stack.push(node);
            if(reverse == true){
                node = node.right;
            }
            else{
                node = node.left;
            }
        }
    }
}

class Solution {
    public boolean findTarget(TreeNode root, int k) {
        
        BSTIterator nextIterator = new BSTIterator(root, false);
        BSTIterator beforeIterator = new BSTIterator(root, true);
        
        int L = nextIterator.next();
        int R = beforeIterator.next();
        
        while(L<R){
            if(L+R == k) return true;
            
            if(L+R > k){
                // reduce R
                R = beforeIterator.next();
            }
            else{
                L = nextIterator.next();
            }
        }
        
        return false;
    }
}

File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.cpp
class BSTIterator{
    stack<TreeNode*> mystack;
    // reverse -> true -> before
    // reverse -> false -> after;
    bool reverse = true;
    
public:
    BSTIterator(TreeNode* root, bool isReverse){
        reverse = isReverse;
        pushAll(root);
    }
    
    bool hasNext(){
        return mystack.size() != 0;
    }
    
    int next(){
        TreeNode* tmpNode = mystack.top();
        mystack.pop();
        
        if(!reverse) pushAll(tmpNode->right);
        else pushAll(tmpNode->left);
        
        return tmpNode->val;
    }
private:
    void pushAll(TreeNode* root){
        
        while(root != NULL){
            mystack.push(root);
            if(reverse == true){
                root = root->right;
            }
            else{
                root = root->left;
            }
        }
    }
};

class Solution {
public:
    bool findTarget(TreeNode* root, int k) {
        
        if(!root) return false;
        
        // next iterator
        BSTIterator l(root, false);
        
        // before iterator
        BSTIterator r(root, true);
        
        int i = l.next();
        int j = r.next();
        
        while(i<j){
            if(i+j == k) return true;
            if(i+j < k) i = l.next();
            else j = r.next();
        }
        
        return false;
    }
};